presto-main/src/main/java/com/facebook/presto/sql/planner/PlannerUtils.java

```diff
@@ -87,4 +123,155 @@ public static VariableReferenceExpression toVariableReference(Expression express
+        return new MapType(keyType, valueType, keyNativeEquals, keyNativeHashCode);
+    }
+
+    public static RowExpression orNullHashCode(RowExpression expression)
+    {
+        checkArgument(BIGINT.equals(expression.getType()), "expression should be BIGINT type");
+        return new SpecialFormExpression(expression.getSourceLocation(), SpecialFormExpression.Form.COALESCE, BIGINT, expression, constant(NULL_HASH_CODE, BIGINT));
+    }
+
+    public static Optional<RowExpression> getHashExpression(FunctionAndTypeManager functionAndTypeManager, List<VariableReferenceExpression> variables)
+    {
+        if (variables.isEmpty()) {
+            return Optional.empty();
+        }
+
+        RowExpression result = constant(INITIAL_HASH_VALUE, BIGINT);
+        for (VariableReferenceExpression variable : variables) {
+            RowExpression hashField = call(functionAndTypeManager, HASH_CODE, BIGINT, variable);
+            hashField = orNullHashCode(hashField);
+            result = call(functionAndTypeManager, "combine_hash", BIGINT, result, hashField);
+        }
+        return Optional.of(result);
+    }
+
+    public static PlanNode projectExpressions(PlanNode source, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator, List<? extends RowExpression> expressions)
+    {
+        Assignments.Builder assignments = Assignments.builder();
+        for (RowExpression expression : expressions) {
+            assignments.put(variableAllocator.newVariable("expr", expression.getType()), expression);
+        }
+
+        return new ProjectNode(
+                source.getSourceLocation(),
+                planNodeIdAllocator.getNextId(),
+                source,
+                assignments.build(),
+                LOCAL);
+    }
+
+    public static PlanNode addProjections(PlanNode source, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator, List<RowExpression> expressions)
+    {
+        Assignments.Builder assignments = Assignments.builder();
+        for (VariableReferenceExpression variableReferenceExpression : source.getOutputVariables()) {
+            assignments.put(variableReferenceExpression, variableReferenceExpression);
+        }
+
+        for (RowExpression expression : expressions) {
+            assignments.put(variableAllocator.newVariable("expr", expression.getType()), expression);
+        }
+
+        return new ProjectNode(
+                source.getSourceLocation(),
+                planNodeIdAllocator.getNextId(),
+                source,
+                assignments.build(),
+                LOCAL);
+    }
+
+    public static PlanNode addAggregation(PlanNode planNode, FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator planNodeIdAllocator, VariableAllocator variableAllocator, String aggregationFunction, Type type, List<VariableReferenceExpression> groupingKeys, VariableReferenceExpression resultVariable, RowExpression... args)
```

Comment:
Can we add line breaks to make it more readable? Similarly for other methods in this class.

Reply:
We don;'t generally do that for method signatures. We do it only for calls. Following the general project style

Reply:
I have seen plenty of examples both ways. Anyways I would leave it to @rschlussel

Reply:
just saw this, but yeah we also do it in method signatures if the line gets to be too long.